### ISSUE TITLE
attrs class inheriting from dataclass

### ISSUE THREAD

**Issue by indigoviolet** [microsoft/pyright#3439]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**

An `attrs` class inheriting from a `dataclass` behaves like a dataclass inheriting from a dataclass -- expecting that parent fields will be present in the `__init__` etc. Is this possibly due to how they are being supported via the dataclass_transform mechanism?

**Expected behavior**
In the code below, `Child` actually takes only one argument, `bar`. 

**Screenshots or Code**
```python

❯ cat main.py
from dataclasses import dataclass, field

import attrs

@dataclass
class Parent:
    foo: int = field(default=2)


@attrs.define
class Child(Parent):
    bar: str

if __name__ == '__main__':
    c = Child('c')
    print(repr(c))%                                                                                                                                                                                                                                                     
```

```
❯ pyright main.py
No configuration file found.
pyproject.toml file found at /home/venky/dev/test-pyright-dataclasses-attrs.
Loading pyproject.toml file at /home/venky/dev/test-pyright-dataclasses-attrs/pyproject.toml
Pyproject file "/home/venky/dev/test-pyright-dataclasses-attrs/pyproject.toml" is missing "[tool.pyright]" section.
stubPath /home/venky/dev/test-pyright-dataclasses-attrs/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
/home/venky/dev/test-pyright-dataclasses-attrs/main.py
  /home/venky/dev/test-pyright-dataclasses-attrs/main.py:12:5 - error: Fields without default values cannot appear after fields with default values
  /home/venky/dev/test-pyright-dataclasses-attrs/main.py:15:9 - error: Argument missing for parameter "bar" (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations 
Completed in 0.674sec

```

```python
>>> help(Child.__init__)
Help on function __init__ in module main:

__init__(self, bar: str) -> None
    Method generated by attrs for class Child.
```
                                                                                                                                                                                                                     
Repo is https://github.com/indigoviolet/test-pyright-dataclasses-attrs

**VS Code extension or command-line**
command line

**Comment by erictraut** [comment]
In the case of a `dataclass` inheriting from a `dataclass`, this code would result in a runtime exception. If you replace the `@attrs.define` with `@dataclass` and then run the code, you will receive a `TypeError: non-default argument 'bar' follows default argument`. This limitation makes sense because the resulting synthesized constructor needs to create a legal function signature that includes one parameter for each field, and parameters with a default value cannot precede parameters without a default value.

It's not clear to me what `@attrs.define` is doing in this circumstance. It looks like it's ignoring the fact that there's a field called `foo` in the parent class when it synthesizes the `__init__` method. If I try to call `Child(foo=3, bar="c")`, I receive a runtime exception `TypeError: Child.__init__() got an unexpected keyword argument 'foo'`.

So I can't tell if this is a bug in attrs or an intentional deviation from the `dataclass` behavior. In either case, this isn't something that pyright will support. We made the deliberate decision not to support non-standard behaviors in third-party libraries. Instead, we pursued the approach of standardizing these behaviors in PEP 681. The behavior of `dataclass_transform` could be extended in the future if there is sufficient demand, but in its current form, it doesn't support the behavior I'm seeing here.

There are a couple of workarounds I can suggest:
1. Avoid intermixing `@dataclass` and `@attrs.define` within a class hierarchy. Stick with one or the other.
2. Remove the default for `foo` in `Parent`.
3. Add a default for `bar` in `Child`.

**Comment by JelleZijlstra** [comment]
Maybe it should be a type checker error if a class and its base class use dataclass transform behavior for different libraries. It's unlikely to work well in practice; I wouldn't expect attrs to recognize that the parent class is a dataclass and adjust its behavior accordingly.

**Comment by indigoviolet** [comment]
> In the case of a `dataclass` inheriting from a `dataclass`, this code would result in a runtime exception. ..., and parameters with a default value cannot precede parameters without a default value.

Yes, I understand this.

> 
> It's not clear to me what `@attrs.define` is doing in this circumstance. It looks like it's ignoring the fact that there's a field called `foo` in the parent class when it synthesizes the `__init__` method. If I try to call `Child(foo=3, bar="c")`, I receive a runtime exception `TypeError: Child.__init__() got an unexpected keyword argument 'foo'`.
> So I can't tell if this is a bug in attrs or an intentional deviation from the `dataclass` behavior.

I don't think it is a bug - in general, attrs won't be able to handle arbitrary parent classes' internals, and that job is left to `super().__init__()` via one of the post/pre/init hooks that it provides. For instance, the following code would create a Child with `foo=3`:

```python
@attrs.define
class Child(Parent):
    bar: str

    def __attrs_post_init__(self):
        super().__init__(3)
```

Is it a valid expectation that 

- attrs and dataclasses should interop in the way pyright is suggesting?
- that attrs classes have knowledge or control over the classes they inherit from?

**Comment by indigoviolet** [comment]
> It's unlikely to work well in practice; I wouldn't expect attrs to recognize that the parent class is a dataclass and adjust its > behavior accordingly.

@JelleZijlstra Why do you say it is unlikely to work well? I wouldn't _want_ attrs to recognize the parent class's implementation details, I would prefer to use attrs' mechanisms to extend the class via `super().__init__()`. Is there something risky about that that I'm missing?

**Comment by erictraut** [comment]
> I don't think it is a bug

If it's not a bug, then it's an unfortunate design choice. I don't know why it wouldn't adopt the same behavior as `dataclass` in this case rather than adopting a subtly different behavior.

In any case, this isn't something that pyright will support, so you'll need to find a workaround.

**Comment by indigoviolet** [comment]
@erictraut - I'm happy to find a workaround, but could you please help me understand the discussion: why do you think it is an unfortunate design choice for attrs to not introspect the implementation of its parent class? 

I can see how it special cases subclassing from another attrs class, which might be the common case, but surely it would be surprising if it also did that for dataclass parents, pydantic parents etc?

**Comment by erictraut** [comment]
I'm saying that it's unfortunate that attrs chose to deviate from the behavior of dataclass in a subtle way like this, especially if the intent was to interop with dataclasses. The dataclass implementation also introspects the fields in its parent class, and its behavior is well documented. Why not adopt the same behavior of dataclass in this case?

**Comment by indigoviolet** [comment]
Thanks for explaining. I won't waste our time any further, except to say that I'm surprised at _"intent to interop with dataclasses"_ --  AFAICT, attrs is a parallel implementation of PEP557-style data classes, but not one that is expected to treat `@dataclass` classes specially (or vice versa). 

I don't know enough about the `dataclass_transform` mechanism to comment very knowledgeably about this, but is it worth considering whether treating all dataclass-like libraries uniformly is the right call? (No response needed).
